### PR TITLE
For AddNode create a new client router

### PR DESCRIPTION
**Pull request by PavelZaytsev**
## Overview

Description:
If a client invalidates a layout in another thread while performing an AddNode in the current thread, there is a possibility of accidentally pruning the router to the node being added from the NodeRouterPool, preventing a client from initiating an AddNodeWorkflow on the server's orchestrator. It is better to create the client router on demand for pings and bootstraps before issuing a workflow request since the node being added is not a part of the layout yet.

Why should this be merged: 
Fixes the defect and potential bug. 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [x] There are no TODOs left in the code
- [x] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [x] Change is covered by automated tests
- [x] Public API has Javadoc

### CODE REVIEW

File: it/src/test/java/org/corfudb/universe/scenario/WriteAfterResetIT.java [hunk 1]
```diff
@@ -118,7 +118,13 @@ private void writeAfterReset(UniverseWorkflow<Fixture<UniverseParams>> wf) throw
                 }
             }).exceptionally(ex -> {
                 // Bootstrap a new 1 node cluster.
-                corfuCluster.bootstrap();
+                try {
+                    corfuCluster.bootstrap();
+                }
+                catch (Exception e) {
+                    throw new RuntimeException(e);
```
Comment (xiaoqin2012) [inline it/src/test/java/org/corfudb/universe/scenario/WriteAfterResetIT.java:125]: Add a long information here. So we know where the exception is from

File: runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java [hunk 2]
```diff
@@ -76,20 +94,34 @@ private static void bootstrapLayoutServer(IClientRouter router, Layout layout)
...
+                })
+                .thenCompose(bootstrapSucceeded -> {
+                    if (bootstrapSucceeded) return CompletableFuture.completedFuture(layout);
+                    return managementClient.getLayout();
+                })
+                .thenAccept(retrievedLayout -> {
+                    if (retrievedLayout.equals(layout)) {
+                        return;
+                    }
+                    throw new AlreadyBootstrappedException();
```
Comment (xiaoqin2012) [inline runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:121]: Add a log information.

Comment (xnull) [reply runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:121]: Actually, if you throw an exception you must NOT add any logs.
Adding more info into an exception is ok, but not adding logs, which causes extra logs that are not needed

File: it/src/test/java/org/corfudb/universe/scenario/WriteAfterResetIT.java (hunk 1 above)
Comment (PavelZaytsev) [reply it/src/test/java/org/corfudb/universe/scenario/WriteAfterResetIT.java:125]: Sounds good

File: runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java (hunk 2 above)
Comment (PavelZaytsev) [reply runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java]: I made an exception more descriptive

**Comment by xiaoqin2012** [comment]
Before fixing the problem, can you write a test to reproduce the issue in https://bugzilla.eng.vmware.com/show_bug.cgi?id=2536425.

File: it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+     * 4) Create a separate router for adding a node, and invoke the add node procedure.
+     *    Keep invalidating a layout every time RPC is called. The add node should succeed.
+     * 5) Verify that the node is present in the new layout.
+     */
+    @Test(timeout = 300000)
+    public void AddNodeDuringInvalidateLayoutTest() {
+        workflow(wf -> {
+
+                    wf.setupDocker(fixture -> {
+                        fixture.getCluster().numNodes(3);
```
Comment (xnull) [inline it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java:51]: it is a 3 node cluster by default, don't need to specify it

File: it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+                }
+        );
+    }
+
+    private void addNodeDuringInvalidateLayout(UniverseWorkflow<Fixture<UniverseParams>> wf) throws Exception {
+        UniverseParams params = wf.getFixture().data();
+        DockerCorfuCluster corfuCluster = wf.getUniverse()
+                .getGroup(params.getGroupParamByIndex(0).getName());
+        Optional<CorfuServer> first = corfuCluster.nodes().values().asList().stream().findFirst();
```
Comment (xnull) [inline it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java:69]: can be simplified:
```
CorfuServer server0 = corfuCluster.getFirstServer();
```

File: runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java [hunk 5]
```diff
@@ -75,21 +69,52 @@ private static void bootstrapLayoutServer(IClientRouter router, Layout layout)
...
+                        CFUtils.unwrap(ex, AlreadyBootstrappedException.class);
+                    } catch (AlreadyBootstrappedException e) {
+                        log.warn("BootstrapUtil: Server {}:{} already bootstrapped.",
+                                router.getHost(), router.getPort());
+                        return false;
+                    }
+                    return true;
+                })
+                .thenCompose(bootstrapSucceeded -> {
+                    if (bootstrapSucceeded) return CompletableFuture.completedFuture(layout);
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:106]: please add curly braces

File: runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java [hunk 6]
```diff
@@ -75,21 +69,52 @@ private static void bootstrapLayoutServer(IClientRouter router, Layout layout)
...
+     * Bootstrap a server with a specific function.
+     *
+     * @param router          Router connected to the node.
+     * @param layout          Layout to bootstrap with.
+     * @param bootstrapFuture A supplier for the function to bootstrap a server.
+     * @param getLayoutFuture A supplier for the function to get a layout.
+     */
+    private static void bootstrap(IClientRouter router, Layout layout,
+                                 Supplier<CompletableFuture<Boolean>> bootstrapFuture,
+                                 Supplier<CompletableFuture<Layout>> getLayoutFuture) {
```
Comment (xnull) [inline runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:92]: let's rename `getLayoutFuture` to something like: `layoutFutureProvider`
same for bootstrapFuture

File: it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java (hunk 3 above)
Comment (PavelZaytsev) [reply it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java:51]: Fixed

File: it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java (hunk 4 above)
Comment (PavelZaytsev) [reply it/src/test/java/org/corfudb/universe/scenario/AddNodeDuringInvalidateLayoutIT.java:69]: Fixed

File: runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java (hunk 5 above)
Comment (PavelZaytsev) [reply runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:106]: fixed

File: runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java (hunk 6 above)
Comment (PavelZaytsev) [reply runtime/src/main/java/org/corfudb/runtime/BootstrapUtil.java:92]: fixed

**Pull request closed without merging**